src/jit/codegencommon.cpp

```diff
@@ -4888,6 +4888,82 @@ void            CodeGen::genFnPrologCalleeRegArgs(regNumber xtraReg,
 #pragma warning(pop)
 #endif
 
+
+/*****************************************************************************
+* If any Vector3 args are on stack and they are not pass-by-ref, the upper 32bits
+* must be cleared to zeroes. The native compiler doesn't clear the upper bits
+* and there is no way to know if the caller is native or not. So, the upper
+* 32 bits of Vector argument on stack are always cleared to zero.
+*/
+#ifdef FEATURE_SIMD
+void    CodeGen::genClearStackVec3ArgUpperBits()
+{
+#ifdef DEBUG
+    if (verbose)
+        printf("*************** In genClearStackVec3ArgUpperBits()\n");
+#endif
+
+#if !defined(_TARGET_XARCH_)
+    return;
+#endif // !_TARGET_XARCH_
```

Comment:
So far we have established the following facts based on @sejongoh's investigation:

1. Native compiler doesn't guarantee that upper 4 bytes of a SIMD Vector3 struct not zero init'ed whether it is passed/returned in a register or on stack.  That would mean PInvoke returned Vector3 value or its out params or RPInvoke passed Vector3 params are not guaranteed to have its upper 4 bytes zero init'ed.
2. This is not an issue on Windows x64 because params are passed by reference and jit generated code reads/writes only 12 bytes from param.  Similarly return value are returned using retBuf, hence jit allocates a temp and zero initializes and passes it by reference..
3. This is an issue on x64 Unix since params are considered passed by value.

Here are my thoughts:

- First this fix is not needed on x64 windows and just on x64 unix.
- This fix alone would not be sufficient on x64 unix due to the reason mentioned in (1) above.  This fix doesn't handle the case of PInvoke returning a Vector3 value in two return registers.  
- As a fix it is sufficient that we zero init upper 4-bytes of PInvoke returned Vector3 value and RPInvoke passed Vector3 params.  Most common case of managed-to-managed calls don't need to incur the penalty of unnecessarily zero initialization, since Jitted code is always ensuring upper 4-bytes of Vector3 are zero.
- As per my understanding VM generates IL stubs for both PInvokes and RPInvokes.  One possible fix to explore is if it is an IL stub then JIT will explicitly zero init upper 4-bytes of Vector3 params before passing it to a call made by IL stub.  Similarly if an IL stub is returning a Vector3 type return value, JIT would explicitly zero init upper 4-bytes before returning it.  This way the penalty is confined to PInvokes and RPInvokes involving Vector3 types, which as per my understanding is less common.